ultra/train.py

```diff
@@ -0,0 +1,155 @@
+        if not os.path.exists(f"{experiment_dir}/spec.pkl"):
+            if not os.path.exists(experiment_dir):
+                os.makedirs(experiment_dir)
+            with open(f"{experiment_dir}/spec.pkl", "wb") as spec_output:
+                dill.dump(spec, spec_output, pickle.HIGHEST_PROTOCOL)
+        while not dones["__all__"]:
+            if episode.get_itr(AGENT_ID) >= 1000000:  # 1M observation break
+                finished = True
+                break
+            evaluation_check(
+                agent=agent,
+                agent_id=AGENT_ID,
+                policy_class=policy_class,
+                episode=episode,
+                **eval_info,
+                **env.info,
+            )
+            action = agent.act(state, explore=True)
+            observations, rewards, dones, infos = env.step({AGENT_ID: action})
+            next_state = observations[AGENT_ID]
+
+            loss_output = agent.step(
+                state=state,
+                action=action,
+                reward=rewards[AGENT_ID],
+                next_state=next_state,
+                done=dones[AGENT_ID],
+            )
+            episode.record_step(
+                agent_id=AGENT_ID,
+                infos=infos,
+                rewards=rewards,
+                total_step=total_step,
+                loss_output=loss_output,
+            )
+            total_step += 1
+            state = next_state
+
+        episode.record_episode()
+        episode.record_tensorboard(agent_id=AGENT_ID)
+        if finished:
+            break
+
+    env.close()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser("intersection-single-agent")
+    parser.add_argument(
+        "--task", help="Tasks available : [0, 1, 2, 3]", type=str, default="1"
+    )
+    parser.add_argument(
+        "--level",
+        help="Tasks available : [easy, medium, hard, no-traffic]",
+        type=str,
+        default="easy",
+    )
+    parser.add_argument(
+        "--episodes", help="number of training episodes", type=int, default=1000000
+    )
+    parser.add_argument(
+        "--timestep", help="environment timestep (sec)", type=float, default=0.1
+    )
+    parser.add_argument(
+        "--headless", help="run without envision", type=bool, default=False
+    )
+    parser.add_argument(
+        "--eval-episodes", help="number of evaluation episodes", type=int, default=200
+    )
+    parser.add_argument(
+        "--eval-rate",
+        help="evaluation rate based on number of observations",
+        type=int,
+        default=10000,
+    )
+
+    parser.add_argument(
+        "--seed", help="environment seed", default=2, type=int,
+    )
+    args = parser.parse_args()
+
+    num_cpus = max(
+        1, psutil.cpu_count(logical=False) - 1
+    )  # remove `logical=False` to use all cpus
+
+    policy_class = "ultra.baselines.sac:sac-v0"
+    # ray_kwargs = default_ray_kwargs(num_cpus=num_cpus, num_gpus=num_gpus)
```

Comment:
Should this line be commented out, or can users set resources and other ray kwargs elsewhere?

Reply:
@AlexLewandowski it wasn't compatible with ray 0.8, so I commented it out